File: app/src/main/java/com/example/todoappyuriko/MyDatabaseHelper.java
// MyDatabaseHelper.java
package com.example.todoappyuriko;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.Nullable;

class MyDatabaseHelper extends SQLiteOpenHelper {

    private Context context;
    private static final String DATABASE_NAME = "ToDoDatabase.db";
    private static final int DATABASE_VERSION = 1;

    private static final String TABLE_NAME = "ToDoTable";
    private static final String COLUMN_ID = "taskID";
    private static final String COLUMN_TITLE = "taskTitle";
    private static final String COLUMN_KETERANGAN = "taskKeterangan";
    private static final String COLUMN_STATUS = "taskStatus";

    public MyDatabaseHelper(@Nullable Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        this.context = context;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String query = "CREATE TABLE " + TABLE_NAME +
                " (" + COLUMN_ID + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
                COLUMN_TITLE + " TEXT, " +
                COLUMN_KETERANGAN + " TEXT, " +
                COLUMN_STATUS + " TEXT);";

        db.execSQL(query);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
        onCreate(db);
    }

    void tambahTask(String taskTitle, String taskKeterangan, String taskStatus) {
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues cv = new ContentValues();

        cv.put(COLUMN_TITLE, taskTitle);
        cv.put(COLUMN_KETERANGAN, taskKeterangan);
        cv.put(COLUMN_STATUS, taskStatus);

        long result = db.insert(TABLE_NAME, null, cv);
        if (result == -1) {
            Toast.makeText(context, "Tambahkan data gagal", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(context, "Tambahkan data berhasil", Toast.LENGTH_SHORT).show();
            Log.d("Database", "Data added: Title = " + taskTitle + ", Keterangan = " + taskKeterangan + ", Status = " + taskStatus);
        }
        db.close();
    }


    Cursor TampilkanSemuaTask() {
        String query = "SELECT * FROM " + TABLE_NAME;
        SQLiteDatabase db = this.getReadableDatabase();
        return  db.rawQuery(query,null);
//        Cursor cursor = null;
//
//        try {
//            cursor = db.rawQuery(query, null);
//            Log.d("Database", "Number of records retrieved: " + cursor.getCount());
//        } catch (Exception e) {
//            Log.e("Database", "Error getting data: " + e.getMessage());
//        }
//
//        return cursor;
    }

}

File: app/src/main/java/com/example/todoapplearning/MyDatabaseHelper.kt
package com.example.todoapplearning

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

class MyDatabaseHelper (context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION){
    companion object{
        private const val DATABASE_NAME = "ToDoDatabase.db"
        private const val DATABASE_VERSION = 1
        private const val TABLE_NAME = "ToDoTable"

        private const val COLUMN_ID = "taskID"
        private const val COLUMN_TITLE = "taskTitle"
        private const val COLUMN_KETERANGAN: String = "taskKeterangan"
        private const val COLUMN_STATUS: String = "taskStatus"
    }

    override fun onCreate(db: SQLiteDatabase?) {
        val createTableQuery = "CREATE TABLE $TABLE_NAME ($COLUMN_ID INTEGER PRIMARY KEY, $COLUMN_TITLE TEXT, $COLUMN_KETERANGAN TEXT, $COLUMN_STATUS TEXT)"
        db?.execSQL(createTableQuery)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        val dropTableQuery = "DROP TABLE IF EXISTS $TABLE_NAME"
        db?.execSQL(dropTableQuery)
        onCreate(db)
    }

    fun addTask(task: TaskModel){
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_TITLE, task.title)
            put(COLUMN_KETERANGAN, task.keterangan)
            put(COLUMN_STATUS, task.status)
        }
        db.insert(TABLE_NAME, null, values)
        db.close()
    }

    fun getAllTask(): List<TaskModel>{
        val taskList = mutableListOf<TaskModel>()
        val db = readableDatabase
        val query = "SELECT * FROM $TABLE_NAME"
        val cursor = db.rawQuery(query, null)

        while (cursor.moveToNext()){
            val id = cursor.getInt(cursor.getColumnIndexOrThrow(COLUMN_ID))
            val title = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_TITLE))
            val keterangan = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_KETERANGAN))
            val status = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_STATUS))

            val task = TaskModel(id, title, keterangan, status)
            taskList.add(task)
        }
        cursor.close()
        db.close()
        return taskList
    }

    fun updateTask(task: TaskModel){
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_TITLE, task.title)
            put(COLUMN_KETERANGAN, task.keterangan)
            put(COLUMN_STATUS, task.status)
        }
        val whereClause = "$COLUMN_ID = ?"
        val whereArgs = arrayOf(task.id.toString())
        db.update(TABLE_NAME, values, whereClause, whereArgs)
        db.close()
    }

    fun getTaskByID(taskID: Int): TaskModel{
        val db = readableDatabase
        val query = "SELECT * FROM $TABLE_NAME WHERE $COLUMN_ID = $taskID"
        val cursor = db.rawQuery(query, null)
        cursor.moveToFirst()

        val id = cursor.getInt(cursor.getColumnIndexOrThrow(COLUMN_ID))
        val title = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_TITLE))
        val keterangan = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_KETERANGAN))
        val status = cursor.getString(cursor.getColumnIndexOrThrow(COLUMN_STATUS))

        cursor.close()
        db.close()
        return TaskModel(id, title, keterangan, status)
    }

    fun deleteTask(taskID: Int){
        val db = writableDatabase
        val whereClause = "$COLUMN_ID =?"
        val whereArgs = arrayOf(taskID.toString())
        db.delete(TABLE_NAME, whereClause, whereArgs)
        db.close()
    }

}